ARCHITECTURE.md

```diff
@@ -0,0 +1,71 @@
+Architecture of the library
+===
+
+    ELF -> Specifications -> Objects -> Links
+
+ELF
+---
+
+BPF is usually produced by using `clang` to compile a subset of C. `clang` outputs
+an ELF file which contains program byte code (aka BPF), but also metadata for
+maps used by the program. The metadata follows the conventions set by `libbpf`
+shipped with the kernel. Certain ELF sections have special meaning
+and contain structures defined by `libbpf`. Newer versions of `clang` emit
+additional metadata in BPF Type Format (aka BTF).
+
+The library aims to be compatible with `libbpf` so that moving from a C toolchain
+to a Go one creates little friction. To that end, the [ELF reader](elf_reader.go)
+is tested against the Linux selftests and avoids introducing custom behaviour
+if possible. [BTF parsing](internal/btf/) lives in a separate internal package.
+
+The output of the ELF reader is a `CollectionSpec` which encodes
+all of the information contained in the ELF in a form that is easy to work with
+in Go.
+
+### BTF
+
+TODO:
+* Explain that kernel and library use BTF for different purposes
+
+Specifications
+---
+
+`CollectionSpec`, `ProgramSpec` and `MapSpec` are blueprints for in-kernel
+objects and contain everything necessary to execute the relevant `bpf(2)`
+syscalls. Since the ELF reader outputs a `CollectionSpec` it's possible to
+modify clang-compiled BPF code, for example to rewrite constants. At the same
+time the [asm](asm/) package provides an assembler that can be used to generate
+`ProgramSpec` on the fly.
+
+Creating a spec should never require any privileges or be restrictedin some way,
+for example by only allowing programs in native endianness. This ensures that
+the library stays flexible.
+
+Objects
+---
+
+`Program` and `Map` are the result of loading specs into the kernel. Sometimes
+loading a spec will fail because the kernel is too old, or a feature is not
+enabled. There are multiple ways the library deals with that:
+
+* Fallback: older kernels don't allowing naming programs and maps. The library
+  automatically detects support for names, and omits them during load if
+  necessary. This works since names a primarily a debug aid.
```

Comment:
```suggestion
  necessary. This works since names a primarily a debug aid.
```
```suggestion
  necessary. This works since name is primarily a debug aid.
```